opentracing/src/main/java/opentracing/SpanPropagator.java

```diff
@@ -0,0 +1,53 @@
+/**
+ * Copyright 2016 The OpenTracing Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License
+ * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+ * or implied. See the License for the specific language governing permissions and limitations under
+ * the License.
+ */
+package opentracing;
+
+import java.util.Map;
+
+/**
+ * Encodes or Decodes a {@linkplain Span} in binary or text formats.
+ *
+ * <p>The toXXX methods are expected to serialize spans into a pair of values representing
+ * separately the span / span identity, and the trace attributes. This is done specifically
+ * for binary protocols that may represent tracing identity in a dedicated fixed-length slot in the
+ * binary message format, so that it can be inspected efficiently by the middleware / routing layers
+ * without parsing the whole message.
+ */
+public interface SpanPropagator {
+
+  /**
+   * Encodes the span into a binary representation of the span's identity and trace
+   * attributes.
+   */
+  byte[] toBinary(Span span);
```

Comment:
(One other idea now that I'm not typing on my :phone:!)

We could alternatively define an iterator interface for propagation... if done carefully, this would make it possible to distinguish between named and positional parameters, "Tracer Fields" vs the user-defined KV pairs, etc. Something very roughly like this:

```
public interface Propagator {
    PropagationFieldIterator propagate(Span span);

    ....
}

public interface PropagationFieldIterator {
    boolean done();
    PropagationFieldIterator next();

    FieldType fieldType();

    // Returns -1 when not defined... but trace_id and span_id could be `0` and `1` respectively, for example.
    int fieldPosition();

    // Always defined, but not necessarily used when `position() >= 0`:
    String fieldKey();

    // Always defined.
    String fieldValue();

    public enum FieldType {
        TRACER_IMPL,
        ANNOTATION
    };
}
```

^^^ please excuse the terrible Java style... just trying to illustrate the concept, I realize the naming is terrible, etc.


Anyway, I kind of like that idea but think the other direction (deserialization) would be messy. But wanted to throw it out there.

Definitely second @yurishkuro request for a diff on the relevant use-case example.